src/ResourceManager/Common/Commands.ResourceManager.Common/RPRegistrationDelegatingHandler.cs

```diff
@@ -0,0 +1,103 @@
+﻿// ----------------------------------------------------------------------------------
+//
+// Copyright Microsoft Corporation
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// ----------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Azure.Common.Authentication;
+using Microsoft.Azure.Common.Authentication.Models;
+using Microsoft.Azure.Management.Resources;
+using Microsoft.Azure.Management.Resources.Models;
+
+namespace Microsoft.Azure.Common.Authentication.Models
+{
+    public class RPRegistrationDelegatingHandler : DelegatingHandler
+    {
+        /// <summary>
+        /// Contains all providers we have attempted to register 
+        /// </summary>
+        private HashSet<string> registeredProviders;
+
+        private Func<ResourceManagementClient> createClient;
+
+        private Action<string> writeVerbose;
+
+
+        public ResourceManagementClient ResourceManagementClient { get; set; }
```

Comment:
I'm not sure we can take this dependency in common - this will mean that any change in resource manager client will require a change in commands.common.  I think, for this, we need to capture the provider registration call.